File: Algorithms/java/0690. Employee Importance/Solution.java
// https://leetcode-cn.com/problems/employee-importance/

/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    int res = 0;
    Map<Integer, Employee> map = new HashMap<>();
    public int getImportance(List<Employee> employees, int id) {
        for(var emp:employees) map.put(emp.id, emp);
        recur(id);
        return res;
    }

    public void recur(int id) {
        var emp = map.get(id);
        res += emp.importance;
        for(var sid:emp.subordinates) recur(sid);
    }
}

File: Algorithms/cpp/0690. Employee Importance/Solution.cpp
// https://leetcode.com/problems/employee-importance/

/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        map<int, Employee*> id2e;
        for(auto e:employees){
            id2e[e->id] = e;
        }
        function<int(int)> recur;
        recur = [&](int eid){
            int ret = id2e[eid]->importance;
            for(auto sid:id2e[eid]->subordinates){
                ret += recur(sid);
            }
            return ret;
        };
        return recur(id);
    }
};